cpp/augment_authority_data_with_translations.cc

```diff
@@ -0,0 +1,219 @@
+        DbResultSet result_set(db_connection->getLastResultSet());
+        std::vector<one_translation_t> translations;
+        while (const DbRow row = result_set.getNextRow()) {
+            // We are not interested in synonym fields
+            if (IsSynonym(row["status"]))
+                continue;
+            one_translation_t translation(row["translation"], row["language_code"], row["origin"], row["status"]);
+            translations.emplace_back(translation);
+        }
+        all_translations->insert(std::make_pair(ppn, translations));
+        translations.clear();
+    }
+    return;
+}
+
+std::string MapLanguageCode(std::string lang_code) {
+    if (lang_code == "deu")
+        return "de";
+    else if (lang_code == "eng")
+        return "en";
+    else if (lang_code == "fra")
+        return "fr";
+    else if (lang_code == "dut")
+        return "nl";
+    else if (lang_code == "ita")
+        return "it";
+    else if (lang_code == "spa")
+        return "es";
+    else if (lang_code == "hant")
+        return "zh-Hant";
+    else if (lang_code == "hans")
+        return "zh-Hans";
+    else
+        Error("Unknown language code " + lang_code);
+}
+
+void InsertTranslation(MarcRecord *const record, char indicator1, char indicator2, std::string term, std::string language_code, std::string status) {
+    Subfields subfields(indicator1, indicator2);
+    subfields.addSubfield('a', term);
+    subfields.addSubfield('9', "L:" + MapLanguageCode(language_code));
+    subfields.addSubfield('9', "Z:" + std::string(IsSynonym(status) ? "VW" : "AF"));
+    subfields.addSubfield('2', "IxTheo");
+    record->insertField("750", subfields);
+}
+
+char DetermineNextFreeIndicator1(MarcRecord *const record, std::vector<size_t> field_indices) {
+    char new_indicator1 = ' ';
+
+    for (auto field_index : field_indices) {
+        Subfields subfields(record->getSubfields(field_index));
+        char indicator1 = subfields.getIndicator1();
+        if (indicator1 == '9')
+            Error("Indicator1 cannot be further incremented");
+        if (indicator1 == ' ')
+            new_indicator1 = '1';
+        else {
+            new_indicator1 = indicator1 >= new_indicator1 ? (indicator1 + 1) : new_indicator1;
+        }
+    }
+
+    return new_indicator1;
+}
+
+size_t GetFieldIndexForExistingTranslation(const MarcRecord *record, const std::vector<size_t> &field_indices, const std::string &language_code) {
+    for (auto field_index : field_indices) {
+        Subfields subfields_present = record->getSubfields(field_index);
+        if (subfields_present.hasSubfieldWithValue('2', "IxTheo") and subfields_present.hasSubfieldWithValue('9', "L:" + MapLanguageCode(language_code)))
+            return field_index;
+    }
+    return MarcRecord::FIELD_NOT_FOUND;
+}
+
+void ProcessRecord(MarcRecord *const record, const std::map<std::string, std::vector<one_translation_t> > &all_translations) {
+    std::string ppn = record->getControlNumber();
```

Comment:
Please initialise variables w/ braces.  Also, I suspect this variable can be const.